lib/handler/mruby.c

```diff
@@ -759,80 +815,37 @@ void h2o_mruby_run_fiber(h2o_mruby_context_t *ctx, mrb_value receiver, mrb_value
-            generator = h2o_mruby_get_generator(mrb, mrb_ary_entry(output, 2));
-            goto GotException;
-        } else if (status == H2O_MRUBY_CALLBACK_ID_CONFIGURING_APP) {
-            mrb_value pending = mrb_ary_new_capa(mrb, 2);
-            mrb_ary_set(mrb, pending, 0, receiver);
-            mrb_ary_set(mrb, pending, 1, input);
-            mrb_ary_push(mrb, ctx->pendings, pending);
-            goto Exit;
-        } else if (status == H2O_MRUBY_CALLBACK_ID_CONFIGURED_APP) {
-            mrb_int i;
-            mrb_int len = RARRAY_LEN(ctx->pendings);
-            for (i = 0; i != len; ++i) {
-                mrb_value pending = mrb_ary_entry(ctx->pendings, i);
-                mrb_value resumer = mrb_ary_entry(pending, 0);
-                mrb_value args = mrb_ary_entry(pending, 1);
-                h2o_mruby_run_fiber(ctx, resumer, args, NULL);
-            }
-            ctx->pendings = mrb_nil_value();
-
-            mrb_value exc = mrb_ary_entry(output, 1);
-            if (!mrb_nil_p(exc)) {
-                mrb->exc = mrb_obj_ptr(exc);
-                goto GotException;
-            }
-
-            goto Exit;
-        }
-
         receiver = mrb_ary_entry(output, 1);
         mrb_value args = mrb_ary_entry(output, 2);
+        int run_again = 0;
 
-        if (mrb_array_p(args)) {
-            int run_again = 0;
-            switch (status) {
-            case H2O_MRUBY_CALLBACK_ID_SEND_CHUNKED_EOS:
-                input = h2o_mruby_send_chunked_eos_callback(ctx, receiver, args, &run_again);
-                break;
-            case H2O_MRUBY_CALLBACK_ID_HTTP_JOIN_RESPONSE:
-                input = h2o_mruby_http_join_response_callback(ctx, receiver, args, &run_again);
-                break;
-            case H2O_MRUBY_CALLBACK_ID_HTTP_FETCH_CHUNK:
-                input = h2o_mruby_http_fetch_chunk_callback(ctx, receiver, args, &run_again);
-                break;
-            case H2O_MRUBY_CALLBACK_ID_SLEEP:
-                input = h2o_mruby_sleep_callback(ctx, receiver, args, &run_again);
-                break;
-            default:
-                input = mrb_exc_new_str_lit(mrb, E_RUNTIME_ERROR, "unexpected callback id sent from rack app");
-                run_again = 1;
-                break;
-            }
-            if (run_again == 0)
-                goto Exit;
-
+        size_t callback_index = -status - 1;
+        if (callback_index >= ctx->shared->callbacks.size) {
+            input = mrb_exc_new_str_lit(mrb, E_RUNTIME_ERROR, "unexpected callback id sent from rack app");
+            run_again = 1;
         } else {
-            input = mrb_exc_new_str_lit(mrb, E_RUNTIME_ERROR, "callback from rack app did not receive an array arg");
+            h2o_mruby_callback callback = ctx->shared->callbacks.entries[callback_index];
+            input = callback(ctx, input, receiver, args, &run_again);
         }
+        if (mrb->exc != NULL)
+            goto GotException;
```

Comment:
Could you please clarify who is responsible for reporting errors?

I assume that it has been the responsibility of `h2o_mruby_run_fiber`. However, with this change, it seems that we'd be calling `handle_exception` in the callback and then also checking here that `mrb->exc` to see if we should call `handle_exception`.

This seems to be an ambiguity in what the responsibilities of a callback are.

Reply:
Basically `h2o_mruby_run_fiber` has the responsibiilty to check and report errors, so the normal callbacks (http_request, sleep, etc..) doesn't need to care about it. But `send_error_callback` is exceptional because it has to receive the generator as an argument.